File: lib/repository/user_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart' as firebase_auth;
import 'package:hoora/model/user_model.dart';

class UserRepository {
  FirebaseFirestore instance = FirebaseFirestore.instance;

  Future<User> getUser() async {
    String userId = firebase_auth.FirebaseAuth.instance.currentUser!.uid;
    QuerySnapshot snapshot = await instance.collection("user").where("userId", isEqualTo: userId).get();
    return User.fromSnapshot(snapshot.docs[0]);
  }

  Future<List<User>> getAllUsers() async {
    QuerySnapshot snapshot = await instance.collection("user").get();
    List<User> users = User.fromSnapshots(snapshot.docs);
    return users;
  }

  Future<bool> isNicknameAvailable(String nickname) async {
    final snapshot = await instance.collection("user").where("nickname", isEqualTo: nickname).get();
    if (snapshot.docs.isNotEmpty) {
      return false;
    }
    return true;
  }

  Future<void> setNickname(String nickname) async {
    User user = await getUser();
    await instance.collection("user").doc(user.id).update({"nickname": nickname});
  }

  String getEmail() {
    return firebase_auth.FirebaseAuth.instance.currentUser!.email!;
  }

  Future<void> updateProfile({
    required String documentId,
    required String nickname,
    required String firstname,
    required String lastname,
    required String city,
    required String country,
    required DateTime? birthday,
    required Gender gender,
  }) async {
    await instance.collection("user").doc(documentId).update({
      "nickname": nickname,
      "firstname": firstname,
      "lastname": lastname,
      "city": city,
      "country": country,
      "birthday": birthday,
      "gender": gender.name,
    });
  }
}

File: firebase/functions/src/repository/user.repository.ts
import * as admin from "firebase-admin";
import {UserEntity} from "../common/entity/user.entity";

export class UserRepository {
  static async getUser(userId: string): Promise<UserEntity> {
    const snapshot = await admin.firestore().collection("user").where("userId", "==", userId).get();
    const user = UserEntity.fromSnapshot(snapshot.docs[0]);
    return user;
  }

  static async getAll(): Promise<UserEntity[]> {
    const snapshot = await admin.firestore().collection("user").get();
    const users = UserEntity.fromSnapshots(snapshot.docs);
    return users;
  }

  /**
 *
 * @param documentId
 * @param gem
 * @param amountOfferUnlocked
 */
  static async updateUserAfterUnlockingOffer(documentId: string, gem: number, amountOfferUnlocked: number): Promise<void> {
    await admin.firestore().collection("user").doc(documentId).update({
      gem: gem,
      amountOfferUnlocked: amountOfferUnlocked,
    });
  }


  /**
   *
   * @param documentId
   * @param gem
   * @param amountSpotValidated
   * @param experience
   */
  static async updateUserAftervalidatingSpot(documentId: string, gem: number, experience: number, amountSpotValidated: number): Promise<void> {
    await admin.firestore().collection("user").doc(documentId).update({
      gem: gem,
      experience: experience,
      amountSpotValidated: amountSpotValidated,
    });
  }

  /**
   *
   * @param documentId
   * @param gem
   * @param experience
   * @param amountCrowdReportCreated
   */
  static async updateUserAfterCreatingReport(documentId: string, gem: number, experience: number, amountCrowdReportCreated: number): Promise<void> {
    await admin.firestore().collection("user").doc(documentId).update({
      gem: gem,
      experience: experience,
      amountCrowdReportCreated: amountCrowdReportCreated,
    });
  }

  /**
   *
   * @param documentId
   * @param gem
   * @param amountDonation
   */
  static async updateUserAfterDonation(documentId: string, gem: number, amountDonation: number): Promise<void> {
    await admin.firestore().collection("user").doc(documentId).update({
      gem: gem,
      amountDonation: amountDonation,
    });
  }
}
